btree/btree.c

```diff
@@ -1421,22 +1690,32 @@ static void ff_find(struct slot *slot, const struct m0_btree_key *key)
 	M0_PRE(key->k_data.ov_vec.v_count[0] == h->ff_ksize);
 	M0_PRE(key->k_data.ov_vec.v_nr == 1);
 
-	while (i + 1 < j) {
+	if (h->ff_used == 0 ||
+	    (memcmp(ff_key(slot->s_node, 0),
+		    key->k_data.ov_buf[0], h->ff_ksize) >= 0)) {
+		slot->s_idx = 0;
+		return;
+	} else if (memcmp(ff_key(slot->s_node, (h->ff_used - 1)),
+			  key->k_data.ov_buf[0], h->ff_ksize) < 0) {
+		slot->s_idx = h->ff_used;
+		return;
+	}
+
+	do {
 		int m    = (i + j) / 2;
 		int diff = memcmp(ff_key(slot->s_node, m),
 				  key->k_data.ov_buf[0], h->ff_ksize);
-		M0_ASSERT(i < m && m < j);
```

Comment:
What's wrong with this? We definitely want to shrink the interval on each iteration.

Reply:
If used == 2, then i = 0, j = 2 on entry, the loop is entered and m = (0+2)/2 = 1.

I guess we understand the required behaviour of ->nt_find() differently. Note that in addition to insertions it should also work for lookups. Let's leave it as it was for now, work out state machine details and then we will know the exact requirements for t->nt_find().